### ISSUE TITLE
NgxFetch: ParseUrl() failed

### ISSUE THREAD

**Issue by creativeprogramming** [pagespeed/ngx_pagespeed#1148]
Hi,

I'm getting many errors like the following in the log file:

`2016/03/10 18:09:50 [error] 18594#18594: [ngx_pagespeed 1.10.33.6-0] NgxFetch: ParseUrl() failed
2016/03/10 18:10:00 [error] 18588#18588: [ngx_pagespeed 1.10.33.6-0] NgxFetch: ParseUrl() failed
2016/03/10 18:11:27 [error] 18594#18594: [ngx_pagespeed 1.10.33.6-0] NgxFetch: ParseUrl() failed
2016/03/10 18:11:28 [error] 18594#18594: [ngx_pagespeed 1.10.33.6-0] NgxFetch: ParseUrl() failed
2016/03/10 18:11:55 [error] 18587#18587: [ngx_pagespeed 1.10.33.6-0] NgxFetch: ParseUrl() failed
2016/03/10 18:12:00 [error] 18594#18594: [ngx_pagespeed 1.10.33.6-0] NgxFetch: ParseUrl() failed
2016/03/10 18:13:46 [error] 18593#18593: [ngx_pagespeed 1.10.33.6-0] NgxFetch: ParseUrl() failed
2016/03/10 18:13:57 [error] 18587#18587: [ngx_pagespeed 1.10.33.6-0] NgxFetch: ParseUrl() failed
`

Is there a way to get more info about it? Can i configure the log level to print the URL that failed to be parsed? If i cannot configure it, also a code hack to make it printing is accepted as a suggestion. I'm trying to debug it

**Comment by oschaaf** [comment]
@creativeprogramming See https://github.com/pagespeed/ngx_pagespeed/pull/1150

**Comment by creativeprogramming** [comment]
Perfect thanks.

Here's what I'm getting now:

`2016/03/11 15:08:31 [error] 13827#13827: [ngx_pagespeed 1.10.33.6-0] NgxFetch: ParseUrl() failed for [https://it/articles/7917-article-sef-url]:host not found`

the url should be: https:/mysite.it/it/articles/7917-article-sef-url, In other words the hostname is stripped from the url.

Can it be due to the fact I'm using **protocol relative URIs** to link resources? e.g. in my html i've 
 `<base href="//mysite.it/it..." `?

And, what  does this issue imply to pagespeed? If I open those URLs failing in the nginx log (adding hostname) I see them optimized by pagespeed, so it seems all fine to me... what lack can have produced this failed parsing so? I mean is not a warning but an 'error' or i can consider it a warning?

**Comment by jeffkaufman** [comment]
This seems really familiar.  I think we have an open bug about this, where people were seeing the host getting stripped out? My memory is that I wasn't able to reproduce it to fix it, but now I'm not finding the bug.

**Comment by creativeprogramming** [comment]
I disabled protocol relative uri, and still getting this stripped. Anyway it seems i get this error only once per URL, before optimization, when pagespeed cache is empty then all goes fine (Note: I periodically delete all files in the cache directory I'm not using .flush, i do this to free up RAM as i've cache in ramfs)

**Comment by creativeprogramming** [comment]
Ok, i'm going deeper and i disabled the NativeFetcher... now i'm getting this error instead:

` [error] 23935#24047: [ngx_pagespeed 1.10.33.6-0] https://mysite.it/articles/7917-article-sef-url (connecting to:it):0:Error status=670005 (No address associated with hostname) serf_connection_create2`

Note that with native fetcher I was getting:

`2016/03/11 15:08:31 [error] 13827#13827: [ngx_pagespeed 1.10.33.6-0] NgxFetch: ParseUrl() failed for [https://it/articles/7917-article-sef-url]:host not found`

The correct URL should be: https://**mysite.it/it**/articles/7917-article-sef-url

So it seems to me that the parser is understanding the first url path token **"/it/"** as the hostname (it's clear from the serf fetcher log) in the native fetcher log instead we got the hostname stripped, but maybe this is due to the same issue: because the dns resolving of just "it" failed then it sound to me like it tried to rebuild a string with messed pieces, where hostname is null due to some wrong parsing into the data object (i'm guessing, i don't understand well C code)...

Anyway i think the issue can be analyzed here for the native fetcher:

https://github.com/pagespeed/ngx_pagespeed/blob/bcb1eb1dec9f0f872330410a3797eacc1b5871fe/src/ngx_url_async_fetcher.cc#L134
http://lxr.nginx.org/source/src/core/ngx_inet.c#L875

Anyway the same parsing issues seems present also in the Apache version, see this case, sounds similar:

http://letmeask.co/foxfeedspro/details/item_21028/magento-pagespeed-cache-images-from-wrong-url-on-hold/

`Pagespeed tries to retrieve the images from the wrong location. Now it search for: http://www.myurl.nl/catalog/product/cache But this should be http://www.myurl.nl/media/catalog/product/cache`

**Comment by creativeprogramming** [comment]
I tried to the debug (modifing the source) it and it seems that is not the ParseUrl() that is stripping the host, but before it, already there's the issue, I put this:

```c++
message_handler->Message(kError, "Adding fetch to queue %s"
                                ,url.c_str());
    NgxFetch* fetch = new NgxFetch(url, async_fetch,
          message_handler, log_);
    ScopedMutex lock(mutex_);
    pending_fetches_.Add(fetch);
```

And what it printed for failing cases? 

Adding fetch to queue https://it/article/path...

In the other cases (most of request) instead i got an already resolved IP

Adding fetch to queue https://123.123.123.123/it/article/path...

So we've an URL without host sometimes (in the other cases we've it already resoved to IP). 

I tried to figure out in the source where in the 'rewrite thread' the URL is set:

```
// It's called in the rewrite thread. All the fetches are started at
  // this function. It will notify the main thread to start the fetch job.
  void NgxUrlAsyncFetcher::Fetch(const GoogleString& url,
                                 MessageHandler* message_handler,
                                 AsyncFetch* async_fetch) {
```

But i don't understood where this is done for the NgxUrlAsyncFetcher::Fetch (maybe I need to recompile PSOL from source)

What I undestood instead is that `ps_determine_host` in those cases is returing always a correct value (always the dns hostname), so nginx headers are always fine, but it seems that this method it's not called r to build the `GoogleString& url`

```
StringPiece ps_determine_host(ngx_http_request_t* r) {
  StringPiece host = str_to_string_piece(r->headers_in.server);
   ps_srv_conf_t* cfg_s = static_cast<ps_srv_conf_t*>(
      ngx_http_get_module_srv_conf(r, ngx_pagespeed));
      cfg_s->server_context->message_handler()->Message(
        kError, "ps_determine_host: %s ", r->headers_in.server.data );
```
 
So I wonder, where the `const GoogleString& url` that is stored to the pending fetches is built and stored? Surely in PSOL, but it sometimes fails to get the host part, if you can give me some hints about compiling PSOL and putting a print in the point where this is done, I can debug it deeper...

**Comment by jeffkaufman** [comment]
This is fixed in 1.10.33.7

**Comment by creativeprogramming** [comment]
Yes i confirm that yesterday i tried 1.10.33.7 (after some days i turned off pagespeed) and now i'm not getting any error. Thanks
